### PR TITLE
 feat(sdk): Sliding sync has a timeout if all lists require a timeout

### PR DESCRIPTION
**Pull request by Hywan**
This patch updates when sliding sync requests have a `timeout`.

Prior to this patch, all requests had a `timeout` query, set to the
`poll_timeout` duration value. However it means: if there is no data
to return, wait `timeout` milliseconds for new data before returning.
This definition is correct. Problem: if the current range of a list
has no data, the server will wait! It means that, in a situation where
there is no update at all, but the client is fetching all rooms batch by
batch, it will wait `poll_timeout` for each batch!

The behaviour described above is absolutely correct. Some server
implementations are less strict though, and we didn't realise our code
was doing that, because the server had some optimisations to ignore the
timeout if the range wasn't covering all the rooms. Nonetheless, a new
server implementation (namely Synapse) is strict, and it confirms we
have a bug here.

This patch then configures a `timeout` if all lists require a `timeout`,
otherwise there is no `timeout`, which is equivalent to `timeout=0`.

---

* Fixes https://github.com/element-hq/element-x-ios/issues/3165#issuecomment-2295700461

### CODE REVIEW

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 1]
```diff
@@ -415,13 +415,18 @@ impl SlidingSync {
     ) -> Result<(http::Request, RequestConfig, OwnedMutexGuard<SlidingSyncPositionMarkers>)> {
         // Collect requests for lists.
         let mut requests_lists = BTreeMap::new();
+        let all_lists_are_fully_loaded;
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/sliding_sync/mod.rs]: Having auxiliary verbs in variable names always feels super weird. How about `fully_loaded_all_lists`?

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 2]
```diff
@@ -466,10 +471,13 @@ impl SlidingSync {
 
         Span::current().record("pos", &pos);
 
+        // Configure the timeout.
+        let timeout = if all_lists_are_fully_loaded { Some(self.inner.poll_timeout) } else { None };
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/sliding_sync/mod.rs]: Can you expand on that comment, to explain why we don't set the `poll_timeout` sometimes? (basically repeating the commit's message in fewer words, ideally)

File: crates/matrix-sdk/src/sliding_sync/list/mod.rs [hunk 3]
```diff
@@ -94,6 +94,12 @@ impl SlidingSyncList {
         self.inner.state.read().unwrap().clone()
     }
 
+    /// Check whether this list requires a [`http::Request::timeout`] value.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/sliding_sync/list/mod.rs:97]: Maybe expand on that comment too, or link to the other comment where we explain why we set the timeout value or not?

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 4]
```diff
@@ -415,18 +415,18 @@ impl SlidingSync {
     ) -> Result<(http::Request, RequestConfig, OwnedMutexGuard<SlidingSyncPositionMarkers>)> {
         // Collect requests for lists.
         let mut requests_lists = BTreeMap::new();
-        let all_lists_are_fully_loaded;
+        let set_timeout;
 
         {
             let lists = self.inner.lists.read().await;
-            let mut fully_loaded = true; // Start at `true` in case there is zero list.
+            let mut require_timeout = true; // Start at `true` in case there is zero list.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/sliding_sync/mod.rs:422]: Why using a temporary variable, btw? Now that `set_timeout` is a short variable name, we could override it in the loop directly, avoiding two lines of code and a cognitive overhead of one (1) temporary binding :grin:

**Review by bnjbvr — APPROVED** [review]
I'd have wished the last 3 commits would be squashed into 59324a63bdb8b48514c0f35f50a30d7ffe9a9814, since they all update the logic introduced in that commit, and having them as separate commits makes the review somewhat harder. LGTM otherwise, assuming my comments will be addressed.

File: crates/matrix-sdk/src/sliding_sync/list/mod.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/list/mod.rs:97]: Yup, indeed!

File: crates/matrix-sdk/src/sliding_sync/mod.rs (hunk 4 above)
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:422]: Yes, you're right. I didn't want to have the outer `set_timeout` to be mutable, but I've refactored this a little bit. Thanks for the notice.

**Pull request merged**